### PR TITLE
Do not load and install ILM policies when in DSL mode

### PR DESCRIPTION
**Pull request by andreidan**
When running in data stream lifecycle only mode skip parsing, loading, and installing ILM policies.
This will not change the current behaviour for stateful deployments (where ILM is available).

Note that the rest of the resources (index templates) would be statically replaced in stateless
(and will contain the equivalent data stream lifecycle configuration), where the 
`data_streams.lifecycle_only.mode` setting will also be registered with value `true`.

### CODE REVIEW

**Comment by andreidan** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by andreidan** [comment]
@elasticmachine update branch

**Comment by dakrone** [comment]
Why do we need two methods for separating the loading and then returning of the policies? Could we instead perhaps move the loading/parsing to the `getLifecyclePolicies()` method and then not call that method if the flag is not set? Calling `super.getLifecyclePolicies()` in certain places seems like an indication that there is a better way to do it.

**Comment by andreidan** [comment]
> Why do we need two methods for separating the loading and then returning of the policies? Could we instead perhaps move the loading/parsing to the getLifecyclePolicies() method and then not call that method if the flag is not set?

Those are great questions. The reason I took the approach of two methods in this PR is because I wanted to maintain the existing behaviour. Namely:
1. In open source if a particular resource is missing we want the node to not start (so the resources loading must be performed in the constructor) 
2. In open source we would iterate over the policies (i.e. we call `getLifecyclePolicies`) on every cluster state change (to check if we must install the policies). I don't think we should be loading the lifecycle policies on every cluster state change as it's I/O in the cluster state applier thread (and they don't change until an Elasticsearch version upgrade - if that - so it'd be a bit wasteful when it comes to CPU cycles)

**Comment by dakrone** [comment]
> In open source we would iterate over the policies (i.e. we call getLifecyclePolicies) on every cluster state change (to check if we must install the policies). I don't think we should be loading the lifecycle policies on every cluster state change as it's I/O in the cluster state applier thread (and they don't change until an Elasticsearch version upgrade - if that - so it'd be a bit wasteful when it comes to CPU cycles)

Should we have the abstract class memoize the policies to prevent this? Would that change the way we implement this?

**Comment by andreidan** [comment]
@elasticmachine update branch

**Comment by andreidan** [comment]
@dakrone 

> Should we have the abstract class memoize the policies to prevent this? Would that change the way we implement this?

I believe you're proposing something along these lines?
```
    protected List<LifecyclePolicy> getLifecyclePolicies() {
        if (lifecyclePolicies == null) {
            lifecyclePolicies = loadLifecycleConfigurations();
        }
        return lifecyclePolicies;
    }
```
(note we'd still need a `load...` method for child classes to indicate what to load)

I don't particularly like this as it forces us to have the `lifecyclePolicies` field  _not_ be `final` anymore (and potentially exposes us to some concurrency issues - i.e. I believe if we go down this route `lifecyclePolicies` would need to be `volatile`) 
Also note that baking too much logic in the parent `getLifecyclePolicies` method is a bit error prone as we have child classes that want to override it to only install the policies if a particular setting is enabled or not so the memoization could be lost ( https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryTemplateRegistry.java#L85 )

Apologies if I misunderstood your suggestion - let's talk offline if that's the case?

If I look back a bit - we started this conversation because some child classes were calling `super.getLifecyclePolicies()`. 
Would making the `lifecyclePolicies` list `protected` and accessible to child classes help here? (that way child implementors could just say `return lifecyclePolicies`) - I've made this change in a [commit](https://github.com/elastic/elasticsearch/pull/98010/commits/af1a2991bf45c5ea985b35fe69c5bd7d36558f3e) in order to visualise it. Let me know what you think.

**Review by dakrone — APPROVED** [review]
This looks good to me, I do wonder though, in `loadLifecycleConfigurations()` we seem to be doing the same thing over and over that looks like:

```java
myListOfPolicyConfigs.stream().map(config -> config.load(LifecyclePolicyConfig.DEFAULT_X_CONTENT_REGISTRY)).toList();
```

What if instead we just returned a `List<LifecyclePolicyConfig>` and had the `IndexTemplateRegistry` do the call to `.load` with the default XContent registry? Then each overriding class would just be returning a list of configs. What do you think?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java [hunk 1]
```diff
@@ -500,8 +540,12 @@ public void onFailure(Exception e) {
     }
 
     private void addIndexLifecyclePoliciesIfMissing(ClusterState state) {
+        if (isDataStreamsLifecycleOnlyMode(state.getMetadata().settings())) {
+            logger.debug("running in data stream lifecycle only mode. skipping the installation of ILM policies.");
```
[the remark this answers is not in the document]
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java]: Sounds good. Made both changes. Thanks for the review.

**Pull request merged**